Add new `ballista-cli` crate

andygrove opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes https://github.com/apache/arrow-datafusion/issues/2493 and https://github.com/apache/arrow-datafusion/issues/2433 and https://github.com/apache/arrow-datafusion/issues/2517

 # Rationale for this change
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

- Allows us to release DataFusion and Ballista separately. This was not practical before because the datafusion-cli depended on ballista which in turn depended on datafusion, leading to issues such as https://github.com/apache/arrow-datafusion/issues/2433)
- Allows DataFusion and Ballista CLIs to evolve separately without Ballista adding development overhead to the DataFusion CLI

Here is the new crate dependency diagram:

![crate-deps](https://user-images.githubusercontent.com/934084/167441480-e7f39c66-6a91-43ee-8506-47297b19bd01.svg)

# What changes are included in this PR?
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

- [x] Create new ballista-cli crate by copying all of datafusion-cli crate
- [x] Remove ballista features from datafusion-cli crate
- [x] Update dependency diagram
- [x] Refactor to have ballista-cli re-use some of the code from datafusion-cli instead of duplicating all of the code
- [x] Update CI scripts
- [x] Update user guide

# Are there any user-facing changes?
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

Yes

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

alamb commented on ballista-cli/src/context.rs:
```diff
@@ -0,0 +1,90 @@
+use std::sync::Arc;
+
+/// The CLI supports using a local DataFusion context or a distributed BallistaContext
+pub enum Context {
```
this can probably be simplified over time as this cli is always for ballista

andygrove replied:
Agreed. I think ballista still needs a local option but it should be using a local (in-process) ballista scheduler and executor and not using datafusion directly.

alamb commented:
Thanks @andygrove  looks much better to me

andygrove commented:
Thanks for the review @alamb. I pushed another commit to fix the `datafusion-cli` lock file that had conflicts after merging from master and also added the `ballista-cli` lock file that was previously ignored. I also updated the CI integration test to build the `ballista-cli`.

andygrove commented:
AMD64 tests failed with `No space left on device` ... trying again

andygrove commented:
AMD64 tests failed with `No space left on device` 3 times in a row now.

andygrove commented:
@alamb All PR builds are failing on AMD64 with `no space left on device`. Any idea what we need to do to resolve?

andygrove commented:
> > 
> 
> @alamb All PR builds are failing on AMD64 with no space left on device. Any idea what we need to do to resolve?
> 
> Perhaps we can try to update the cache-key in https://github.com/apache/arrow-datafusion/blob/master/.github/workflows/rust.yml#L55 to see if something about the old caches is too large?

Thanks. I suppose this cache just continues to grow forever and doesn't clear out old versions of deps that we no longer use? Perhaps we should include the month name in the cache so it resets periodically?

andygrove commented:
@alamb Moving the ballista tests to a separate job fixes this. Ok to merge?

andygrove commented on .github/workflows/rust.yml:
```diff
@@ -123,9 +115,9 @@ jobs:
         run: |
           export ARROW_TEST_DATA=$(pwd)/testing/data
           export PARQUET_TEST_DATA=$(pwd)/parquet-testing/data
+          cargo test
```
We were previously only running `cargo test` in the `datafusion-examples` crate in this step

andygrove commented on .github/workflows/rust.yml:
```diff
@@ -75,14 +75,6 @@ jobs:
           CARGO_TARGET_DIR: "/github/home/target"
-      - name: Check DataFusion CLI builds with ballista
-        run: |
-          (cd datafusion-cli && cargo check)
-          (cd datafusion-cli && cargo check --no-default-features)
-          (cd datafusion-cli && cargo check --features=ballista)
```
We test `datafusion-cli` in the integration tests and we no longer support running this with ballista
